alea/include/alps/alea/core.hpp

```diff
@@ -37,39 +37,95 @@ template <typename T>
 struct traits;
 
 /**
- * Data sink as a thin wrapper around a continuous array.
+ * Data view as a thin wrapper around a continuous array.
  *
- * Basically collects a pointer continuous array (vector, Eigen array, C
- * array etc.) together with its size.  This is sufficiently generic for the
- * virtual interface of `estimator`.
+ * Basically collects a pointer continuous array together with its size.  Note
+ * that the view does not own the `data` pointer.
  */
 template <typename T>
-class sink
+class view
 {
 public:
     typedef T value_type;
 
 public:
     /** Construct view on nothing */
-    sink() : data_(nullptr), size_(0) { }
+    view() : data_(nullptr), size_(0) { }
 
     /** Construct view on data area with size */
-    sink(T *data, size_t size) : data_(data), size_(size) { }
+    view(T *data, size_t size)
+        : data_(data), size_(size)
+    { }
 
     /** Get pointer to zeroth element */
     T *data() { return data_; }
 
     /** Get pointer to zeroth element */
     const T *data() const { return data_; }
 
-    /** Get size of data space */
+    /** Return size */
     size_t size() const { return size_; }
 
 private:
     T *data_;
     size_t size_;
 };
 
+/**
+ * Data view as a thin wrapper around a continuous multi-dimensional array.
+ *
+ * Basically collects a pointer continuous array in ROW-MAJOR format (vector,
+ * transposed Eigen array, C array etc.) together with its shape.  Note that
+ * the view neither owns the `data` pointer nor the `shape` pointer.  Thus it
+ * amends `view<T>` with shape information.
+ */
+template <typename T>
+class ndview
+    : public view<T>
+{
+public:
+    typedef T value_type;
+
+public:
+    /** Construct view on nothing */
+    ndview() : view<T>(), shape_(nullptr), ndim_(0) { }
+
+    /** Construct view on data area with shape */
+    ndview(T *data, const size_t *shape, size_t ndim)
```

Comment:
const size_t * shape? I don't like pointer here.
why not array or vector?:

    template<size_t N>
    void compute(T *data, const std::array<size_t, N> & shape, size_t ndim) {

Reply:
Well, essentially because of performance consideration (I wanted to avoid unnecessary malloc's and the number of dimensions is not fixed at compile time). The main point is that there is an intrinsic scoping issue here anyway, because ndview, similar to, say, Eigen::Map, does not own its data pointer. Thus, while I can make the shape a std::vector, it does not make the scoping issue go away.

Reply:
the data pointer is 100% ok, but shape?
It will make life at least easy, because, finally many people will declare shape as something like a vector and then pass as .data(), .size(). 

another question, am I wrong, that you can't return view from the function if you construct shape inside this function? something like this:

     ndview create(blah blah blah) {
         std::vector<size_t> shape{{1,2,3,4}};
         return ndview(data, shape.data(), shape.size());
     }